File: java/sqlbuilder/Selection.java
package com.clickability.cms.dataaccess.sqlbuilder;

import java.util.StringJoiner;

public class Selection implements Expression {

	private StringJoiner selectList;
	private Resource base;
	private int size;
	
	public Selection() {
		this.selectList = new StringJoiner(",");
		this.size = 0;
	}
	
	public Selection(Resource base) {
		this();
		this.base = base;
	}
	
	public Selection(String ...columns) {
		this();
		for (String column : columns) {
			this.append(column);
		}
	}
	
	public Selection(Column ...columns) {
		this();
		for (Column column : columns) {
			this.append(column.get());
		}
	}
	
	public int size() {
		return this.size;
	}
	
	protected void append(String item) {
		this.size += 1;
		this.selectList.add(item);
	}
	
	public Selection add(String...columns) {
		for (String column : columns) {
			this.append(column);
		}
		return this;
	}

	public Selection add(Column...columns) {
		for (Column column : columns) {
			this.append(column.get());
		}
		return this;
	}

	@Override
	public String get() {
		String selection = "";
		if (this.base != null) {
			selection += this.base.alias();
		}
		if (this.size > 0) {
			selection += this.selectList.toString();
		}
		return  selection;
	}

}

File: java/queryBuilder/Selection.kt
package com.clickability.cms.dataaccess.sqlbuilder

import java.util.StringJoiner

class Selection() : Expression {

    private val selectList: StringJoiner
    private val base: Resource?
    private var size: Int = 0

    init {
        this.selectList = StringJoiner(",")
        this.size = 0
    }

    constructor(base: Resource) : this() {
        this.base = base
    }

    constructor(vararg columns: String) : this() {
        for (column in columns) {
            this.append(column)
        }
    }

    constructor(vararg columns: Column) : this() {
        for (column in columns) {
            this.append(column.get())
        }
    }

    fun size(): Int {
        return this.size
    }

    protected fun append(item: String) {
        this.size += 1
        this.selectList.add(item)
    }

    fun add(vararg columns: String): Selection {
        for (column in columns) {
            this.append(column)
        }
        return this
    }

    fun add(vararg columns: Column): Selection {
        for (column in columns) {
            this.append(column.get())
        }
        return this
    }

    @Override
    fun get(): String {
        var selection = ""
        if (this.base != null) {
            selection += this.base!!.alias()
        }
        if (this.size > 0) {
            selection += this.selectList.toString()
        }
        return selection
    }

}
